Increase LayoutTestFinder test coverage

gsnedders opened the pull request:
#### 24f491ae94894b6c52376c9efb9b76ee817c5a9b
<pre>
Increase LayoutTestFinder test coverage
<a href="https://bugs.webkit.org/show_bug.cgi?id=268524">https://bugs.webkit.org/show_bug.cgi?id=268524</a>

Reviewed by Jonathan Bedard.

As part of bug 220421, rewriting LayoutTestFinder, many new tests have
been written. However, these can stand on their own as a useful increase
in test coverage.

With this, a `with_expectations` argument is added to
LayoutTestFinder.find_tests which fully populates the Test object,
similar to the new implementation. This is not enabled by default as it
is very slow, having to read each directory many times.

This also makes some minor changes to ensure the tests reliably pass, at
least on non-Windows platforms. (Some of the LayoutTestFinder tests
already fail on Windows; these new tests continue to demonstrate
differences, which are almost all bugs in our current implementation.)

* Tools/Scripts/webkitpy/common/system/filesystem.py:
(FileSystem.__init__): Avoid storing os.sep/os.pardir at import time, so that pyfakefs can rewrite them.
(FileSystem.remove): Ditto, but os.remove.
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py:
(LayoutTestFinder.find_tests): Add a with_expectations keyword argument that fully populates the Test object.
(LayoutTestFinder.find_tests_by_path): Ditto.
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy_unittest.py:
(LayoutTestFinderTests.setUp):
(LayoutTestFinderTests.test_supported_test_extensions):
(LayoutTestFinderTests.test_includes_other_platforms_fallback):
(LayoutTestFinderTests.test_find_platform):
(LayoutTestFinderTests.test_find_platform_self):
(LayoutTestFinderTests.test_find_overridden):
(LayoutTestFinderTests):
(LayoutTestFinderTests.test_find_overridden_default):
(LayoutTestFinderTests.test_find_overridden_platform_self):
(LayoutTestFinderTests.test_find_overridden_platform_other):
* Tools/Scripts/webkitpy/layout_tests/models/test.py:
(Test): Fix defaults to be the correct types.
* Tools/Scripts/webkitpy/port/base.py:
(Port): Use a tuple, rather than a set, to give consistent enumeration order over _supported_reference_extensions.
* Tools/Scripts/webkitpy/port/test.py:

Canonical link: <a href="https://commits.webkit.org/273913@main">https://commits.webkit.org/273913@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/537a0c8cdafbafbb460474b7e3c92d469910ec60

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/37031 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/15942 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/39337 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/39613 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/33095 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/38321 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/18563 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/13074 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/31611 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/37592 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/13494 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/32647 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/11718 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/37114 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/11769 "Passed tests") | [❌ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/32944 "Found 2 new test failures: fast/images/avif-as-image.html, fast/images/avif-heif-container-as-image.html (failure)") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/40868 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/33588 "Passed tests") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/34391 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/37633 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/12042 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/9843 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/35766 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/13687 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/12417 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4813 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/12951 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

JonWBedard reviewed (approved):
Feel like we need to re-evaluate if the behavior of each of these test cases is actually correct once you land the re-write...many of these are very surprising to me.

gsnedders committed:
Increase LayoutTestFinder test coverage
https://bugs.webkit.org/show_bug.cgi?id=268524

Reviewed by Jonathan Bedard.

As part of bug 220421, rewriting LayoutTestFinder, many new tests have
been written. However, these can stand on their own as a useful increase
in test coverage.

With this, a `with_expectations` argument is added to
LayoutTestFinder.find_tests which fully populates the Test object,
similar to the new implementation. This is not enabled by default as it
is very slow, having to read each directory many times.

This also makes some minor changes to ensure the tests reliably pass, at
least on non-Windows platforms. (Some of the LayoutTestFinder tests
already fail on Windows; these new tests continue to demonstrate
differences, which are almost all bugs in our current implementation.)

* Tools/Scripts/webkitpy/common/system/filesystem.py:
(FileSystem.__init__): Avoid storing os.sep/os.pardir at import time, so that pyfakefs can rewrite them.
(FileSystem.remove): Ditto, but os.remove.
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy.py:
(LayoutTestFinder.find_tests): Add a with_expectations keyword argument that fully populates the Test object.
(LayoutTestFinder.find_tests_by_path): Ditto.
* Tools/Scripts/webkitpy/layout_tests/controllers/layout_test_finder_legacy_unittest.py:
(LayoutTestFinderTests.setUp):
(LayoutTestFinderTests.test_supported_test_extensions):
(LayoutTestFinderTests.test_includes_other_platforms_fallback):
(LayoutTestFinderTests.test_find_platform):
(LayoutTestFinderTests.test_find_platform_self):
(LayoutTestFinderTests.test_find_overridden):
(LayoutTestFinderTests):
(LayoutTestFinderTests.test_find_overridden_default):
(LayoutTestFinderTests.test_find_overridden_platform_self):
(LayoutTestFinderTests.test_find_overridden_platform_other):
* Tools/Scripts/webkitpy/layout_tests/models/test.py:
(Test): Fix defaults to be the correct types.
* Tools/Scripts/webkitpy/port/base.py:
(Port): Use a tuple, rather than a set, to give consistent enumeration order over _supported_reference_extensions.
* Tools/Scripts/webkitpy/port/test.py:

Canonical link: https://commits.webkit.org/273913@main
